lib/datastore/transaction.js

```diff
@@ -375,9 +376,11 @@ Transaction.prototype.commit_ = function(callback) {
       }
     })
 
-    // Group entities together by action (delete or save).
+    // Group entities together by method: `save` mutations, then `delete`. Note:
+    // `save` mutations being first is required to maintain order when assigning
+    // IDs to incomplete keys.
     .sort(function(a, b) {
-      return a.method > b.method ? 1 : a.method < b.method ? -1 : 0;
+      return a.method < b.method ? 1 : a.method > b.method ? -1 : 0;
```

Comment:
Found a bug here. This groups methods by their name (`save` or `delete`). Previously, it was just going alphabetical, but I switched it so that `saves` come first. Important for the reasons explained in the comment above.